LibRZ/src/Elements/ConicLens.cpp

```diff
@@ -96,81 +105,72 @@ ConicLens::recalcModel()
 
   dZ[0] = dZ[1] = .5 * m_thickness;
   
-  // CHECK OFFSET IS WITHIN EXPECTED VALUE
-  //double maxOffset = ((m_displacement[1] + m_displacement[2]) * (m_displacement[1] + m_displacement[2])) / ((m_K[1] + m_K[2]) * (m_K[1] + m_K[2]));
-  //double r_max = ((m_K[1] + m_K[2]) * (m_K[1] + m_K[2])) / ((m_displacement[1] + m_displacement[2]) * (m_displacement[1] + m_displacement[2]));
-  //double r_max = ((m_thickness * m_thickness) / ((m_K[1] + m_K[2]) * (m_K[1] + m_K[2]))) / 2;
-  double maxOffset = sqrt(m_radius * m_radius - 0.25 * (m_thickness + m_displacement[1] + m_displacement[2]) * (m_thickness + m_displacement[1] + m_displacement[2]));
-  double offset = std::sqrt(m_x0 * m_x0 + m_y0 * m_y0);
+  Real Rmax = RZ::ConicSurface::Rmax(1 * sigma[0], Rc[0], m_K[0], m_displacement[0], -1 * sigma[1], Rc[1], m_K[1], m_displacement[1], m_thickness);
   
-  if (offset > maxOffset) {
-    double factor = maxOffset / offset;
-    m_x0 *= factor;
-    m_y0 *= factor;
-  }
-  //
-
-  // Input focal plane: located at -f minus half the thickness
-  m_frontFocalPlane->setDistance(+(dZ[0] + m_focalLength[0])* Vec3::eZ());
-  m_objectPlane->setDistance(+(dZ[0] + 2 * m_focalLength[0]) * Vec3::eZ());
-
-  m_inputBoundary->setRadius(m_radius);
-  m_inputBoundary->setCurvatureRadius(Rc[0]);
-  m_inputBoundary->setMedia(nullptr, &m_glass);
-  m_inputBoundary->setConicConstant(m_K[0]);
-  m_inputBoundary->setConvex(convex[0]);
-  m_inputBoundary->setCenterOffset(m_x0, m_y0);
-
-  m_frontCap.setRadius(m_radius);
-  m_frontCap.setCurvatureRadius(Rc[0]);
-  m_frontCap.setConicConstant(m_K[0]);
-  m_frontCap.setConvex(convex[0]);
-  m_frontCap.setInvertNormals(false);
-  m_frontCap.setCenterOffset(m_x0, m_y0);
-  m_frontCap.requestRecalc();
-
-  // Output focal plane: opposite side
-  m_backFocalPlane->setDistance(-(dZ[1] + m_focalLength[1]) * Vec3::eZ());
-  m_imagePlane->setDistance(-(dZ[1] + 2 * m_focalLength[1]) * Vec3::eZ());
-
-  m_outputBoundary->setRadius(m_radius);
-  m_outputBoundary->setCurvatureRadius(Rc[1]);
-  m_outputBoundary->setMedia(&m_glass, nullptr);
-  m_outputBoundary->setConicConstant(m_K[1]);
-  m_outputBoundary->setCenterOffset(m_x0, m_y0);
-  m_outputBoundary->setConvex(!convex[1]);
+  if (sqrt(m_x0 * m_x0 + m_y0 * m_y0) + m_radius > Rmax) {
```

Comment:
Performance: try to avoid square roots whenever you can. While it is not critical here, do something like:

```c
if (m_x0 * m_x0 + m_y0 * m_y0 > (Rmax - m_radius) * (Rmax - m_radius)) {
  RZWarning("...");
```